File: algorithms/merge_sort/merge_sort.py
from random import randint


def merge_sort(arr: list[int], lbound: int, ubound: int):
    if ubound - lbound == 0:
        return

    midpoint = int((lbound + ubound) / 2)

    merge_sort(arr, lbound, midpoint)
    merge_sort(arr, midpoint + 1, ubound)
    merge(arr, lbound, midpoint, ubound)


def merge(arr: list[int], lbound: int, midpoint: int, ubound: int):
    
    flen = midpoint - lbound + 1
    rlen = ubound - midpoint

    fsubarr = [arr[lbound + i] for i in range(flen)]
    rsubarr = [arr[midpoint + 1 + j] for j in range(rlen)]

    i = 0
    j = 0
    k = lbound

    while i < flen and j < rlen:
        if (fsubarr[i] <= rsubarr[j]):
            arr[k] = fsubarr[i]
            i += 1
        else:
            arr[k] = rsubarr[j]
            j += 1
        k += 1

    while i < flen:
        arr[k] = fsubarr[i]
        i += 1
        k += 1

    while j < rlen:
        arr[k] = rsubarr[j]
        j += 1
        k += 1


def main():

    arr = [randint(1, 100) for _ in range(10)]

    print("Original array:")
    print(arr)

    merge_sort(arr, 0, len(arr) - 1)

    print("New array")
    print(arr)


if __name__ == "__main__":
    main()

File: algorithms/merge_sort/merge_sort.c
#include<stdio.h>
#include<stdlib.h>


void merge(int *arr, int lbound, int midpoint, int ubound);


void merge_sort(int *arr, int lbound, int ubound) {
    if (lbound == ubound)
        return;

    int midpoint = (lbound + ubound) / 2;
    merge_sort(arr, lbound, midpoint);
    merge_sort(arr, midpoint + 1, ubound);
    merge(arr, lbound, midpoint, ubound);
}


void merge(int *arr, int lbound, int midpoint, int ubound) {
    int flen = midpoint - lbound + 1;  // length of left-side subarray
    int rlen = ubound - midpoint;      // length of right-side subarray
    int fsubarr[flen], rsubarr[rlen];

    int i, j = 0;
    for (i = 0; i < flen; i++) {
        fsubarr[i] = arr[lbound + i];
    }

    for (j = 0; j < rlen; j++) {
        rsubarr[j] = arr[midpoint + 1 + j];
    }

    i = 0;
    j = 0;
    int k = lbound;

    while (i < flen && j < rlen) {
        if (fsubarr[i] <= rsubarr[j]) {
            arr[k] = fsubarr[i];
            i++;
        } else {
            arr[k] = rsubarr[j];
            j++;
        }
        k++;
    }

    while (i < flen) {
        arr[k] = fsubarr[i];
        i++;
        k++;
    }

    while (j < rlen) {
        arr[k] = rsubarr[j];
        j++;
        k++;
    }
}


int main(int argc, char *argv[]) {
    int n = argc - 1;
    int *arr = malloc(n * sizeof(int));

    printf("Original array:\n");
    for (int i = 0; i < n; i++) {
        char *endPtr;
        arr[i] = strtol(argv[i + 1], &endPtr, 10);
        printf("%d\t", arr[i]);
    }

    merge_sort(arr, 0, n - 1);

    printf("\nNew array is:\n");
    for (int i = 0; i < n; i++) {
        printf("%d\t", arr[i]);
    }
}
